### DISCUSSION TITLE
Query taking 40-45 seconds to execute

### DISCUSSION THREAD

**Discussion by abhinav99ash29** [JanusGraph/janusgraph#3883] [Q&A] [answered]
Hey community, 
I have a use case where starting from a particular vertex I need to output it's ancestors having a particular type, something like this

```g.V(<id>).emit().repeat(__.in("label1", "label2").dedup().simplePath()).has("type", "<xyz>")```

The graph has around 25k vertices that are emitted from this traversal and after filtering via has steps around 4k are left.

I have used ```dedup``` in repeat step because some of the traversals point to the same vertices, also the graph has many levels from the current node to the top level ancestors.

Without the has step the query performance is good, but when adding the has step in the query, the query time increase by a lot.

Please let us know the right approach to make these type of query where starting from a vertex we need to find all it's parents of a specific type.

**Comment by porunov** [comment, marked as answer]
Hi, `has` step in your query doesn't apply multi-query optimization. Please, upgrade to the latest JanusGraph commit version `1.0.0-20230705-152651.16846ad` to use those optimizations.
You can find distribution builds here: https://github.com/JanusGraph/janusgraph/actions/runs/5466111451 (scroll down and click on `distribution-builds`).
The optimization is not yet officially released, but should be released in `1.0.0-rc3` or the final `1.0.0` release.
Also, we plan to add additional optimizations specifically for CQL and HBase storage backends into `1.0.0` release (not yet merged).
Hope it helps.

**Reply by abhinav99ash29** [reply]
Hey @porunov , thanks for the reply.
We will try to upgrade and check the query performance.

I wanted to confirm that for my use case of fetching all ancestors of a type starting from a vertex, is the gremlin query written good or are there any other steps we can use to optimise it?

**Reply by porunov** [reply]
`simplePath` is a tricky step and `dedup` might be unnecessary if you rewrite your query. I would use something like below instead of your query:
```
g.withSideEffect("v", new HashSet<>()) // add side effect with a hash set where we will collect each visited vertex
        .V(<id>).aggregate(Scope.local, "v") // mark your first vertex as visited
        .emit().repeat(__.in("label1", "label2").where(P.without("v")).aggregate(Scope.local, "v")) // fetch all available ancestors skipping already traversed vertices
        .has("type", "<xyz>").toList(); // filter all returned vertices by `<xyz>` type
```

Notice, this query (as well as the query in your question) traverses all ancestors until no more vertices can be fetched by incoming `label1` or `label2` edges. If you don't want to continue traversing the branch after you reached the ancestor with the specific type you can remove the last `has` step and replace `emit()` with `until(__.has("type", "<xyz>"))`.

The upgrade to `1.0.0-20230705-152651.16846ad` or later is still needed in your case to ensure `has` step enables proper optimizations. 

Notice, usually it's better to ask Gremlin specific questions in Gremlin-users groups as they may assist you better with your queries. (See [TinkerPop Discord](https://discord.com/invite/kfebyvDM9H), [TinkerPop Google Group](https://groups.google.com/g/gremlin-users), [TinkerPop Documentation](https://tinkerpop.apache.org/docs/current/reference/)).

**Reply by abhinav99ash29** [reply]
Thanks for the thorough explanation @porunov.
We will try to update our query and see if there is a performance increase.

Is there any expected release date for the version 1.0.0?

**Reply by porunov** [reply]
We don’t have a date set for the release. 
I would say most likely by the end of the August or early September, but that can be changed depending on the release process.

**Reply by abhinav99ash29** [reply]
Got it, thanks for the info @porunov
